### PR TITLE
Selinux rules

### LINKED ISSUE #62: Role freezes waiting for signatures on rhel-like OSes.
# 🐛 Summary #

While deploying, the role freezes on this step:

```
TASK [cisagov.clamav : Wait for new signatures to be downloaded and installed by freshclam] *********************************************************************************************************************
```

## To reproduce ##

1. Use a SELinux enabled host as your ansible target (I could reproduce it with Cent OS 7, RHEL 8)
2. Host a clamav private mirror on TCP port 8000 (as it is [the one cvdupdate uses by default](https://github.com/Cisco-Talos/cvdupdate))
3. Declare the configuration below in ansible:

```
clamav_freshclam_configuration:
  DatabaseMirror: ['db.local.clamav.net', 'database.clamav.net']
  Bytecode: 'true'
  PrivateMirror: "http://{{ clamav_private_mirror }}:8000"
  ScriptedUpdates: 'no'
  HTTPProxyServer:
  HTTPProxyPort:
  HTTPProxyUsername:
  HTTPProxyPassword:
```

## Expected behvior ##

Role deployment is successful.

## Any helpful log output or screenshots ##

While the role is stalled, running `systemctl status clamav-freshclam` on the host gives the following output:

```
déc. 05 14:22:47 test-centos systemd[1]: Started ClamAV virus database updater.
déc. 05 14:22:47 test-centos freshclam[18474]: ClamAV update process started at Mon Dec  5 14:22:47 2022
déc. 05 14:22:47 test-centos freshclam[18474]: WARNING: DNS Update Info disabled. Falling back to HTTP mode.
déc. 05 14:22:47 test-centos freshclam[18474]: Trying to retrieve CVD header from http://cvdupdate:8000/daily.cld
déc. 05 14:22:47 test-centos freshclam[18474]: WARNING: remote_cvdhead: Download failed (7) WARNING:  Message: Couldn't connect to server
déc. 05 14:22:47 test-centos freshclam[18474]: Trying to retrieve CVD header from http://cvdupdate:8000/daily.cvd
déc. 05 14:22:47 test-centos freshclam[18474]: WARNING: remote_cvdhead: Download failed (7) WARNING:  Message: Couldn't connect to server
déc. 05 14:22:47 test-centos freshclam[18474]: WARNING: Failed to get daily database version information from server: http://cvdupdate:8000
déc. 05 14:22:47 test-centos freshclam[18474]: ERROR: check_for_new_database_version: Failed to find daily database using server http://cvdupdate:8000.
déc. 05 14:22:47 test-centos freshclam[18474]: Trying again in 5 secs...
```

Audit logs show the following:

```
type=AVC msg=audit(1671773505.429:258326): avc:  denied  { name_connect } for  pid=1068 comm="freshclam" dest=8000 scontext=system_u:system_r:antivirus_t:s0 tcontext=system_u:object_r:soundd_port_t:s0 tclass=tcp_socket permissive=1
type=SYSCALL msg=audit(1671773505.429:258326): arch=c000003e syscall=42 success=no exit=-115 a0=5 a1=7fff36ad6fd0 a2=10 a3=2 items=0 ppid=1 pid=1068 auid=4294967295 uid=992 gid=990 euid=992 suid=992 fsuid=992 egid=990 sgid=990 fsgid=990 tty=(none) ses=4294967295 comm="freshclam" exe="/usr/bin/freshclam" subj=system_u:system_r:antivirus_t:s0 key=(null)
```

Opening a shell on the host to manually run the `freshclam` command as root allows to work around.

### PR DESCRIPTION
**Pull request by mjourdan**
#  #

## 🗣 Description ##

This PR installs SELinux authorization.

## 💭 Motivation and context ##

This closes #62 

## 🧪 Testing ##

Following the procedure given in #62 .

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [ ] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [ ] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [ ] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [ ] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [ ] Tests have been added and/or modified to cover the changes in this PR.
- [ ] All new and existing tests pass.

### CODE REVIEW

File: tasks/main.yml [hunk 1]
```diff
@@ -67,6 +67,24 @@
         state: touch
       when: ansible_distribution == "Fedora"
 
+- name: selinux
+  block:
+
```
Comment (jsf9k) [inline tasks/main.yml]: Remove needless blank line:
```suggestion
```

File: tasks/main.yml [hunk 2]
```diff
@@ -67,6 +67,24 @@
...
+  - name: Generate selinux policy for freshclam
+    ansible.builtin.shell: grep freshclam /var/log/audit/audit.log | audit2allow -M freshclam
+
+  - name: Apply selinux policy for freshclam
+    ansible.builtin.shell: semodule -i freshclam.pp
```
Comment (jsf9k) [inline tasks/main.yml]: Prefer long form options for readability:
```suggestion
    ansible.builtin.shell: semodule --install freshclam.pp
```

**Review by jsf9k — CHANGES_REQUESTED** [review]
Here are a few minor things to get started.  I will look into the failing tests.

File: tasks/main.yml [hunk 3]
```diff
@@ -67,6 +67,24 @@
         state: touch
       when: ansible_distribution == "Fedora"
 
+- name: selinux
```
Comment (jsf9k) [inline tasks/main.yml]: The name of this block should be improved.  How about `Configure SELinux to allow freshclam downloads`?

File: tasks/main.yml [hunk 4]
```diff
@@ -67,6 +67,24 @@
...
+  block:
+
+  - name: Generate selinux policy for freshclam
```
Comment (jsf9k) [inline tasks/main.yml]:
```suggestion
  - name: Generate SELinux policy for freshclam
```

File: tasks/main.yml [hunk 5]
```diff
@@ -67,6 +67,24 @@
...
+    ansible.builtin.shell: grep freshclam /var/log/audit/audit.log | audit2allow -M freshclam
+
+  - name: Apply selinux policy for freshclam
```
Comment (jsf9k) [inline tasks/main.yml]:
```suggestion
  - name: Apply SELinux policy for freshclam
```

File: tasks/main.yml [hunk 6]
```diff
@@ -67,6 +67,24 @@
...
+    ansible.builtin.shell: semodule -i freshclam.pp
+
+  - name: Set selinux antivirus_can_scan_system flag persistently
```
Comment (jsf9k) [inline tasks/main.yml]:
```suggestion
  - name: Set SELinux antivirus_can_scan_system flag persistently
```

File: tasks/main.yml [hunk 7]
```diff
@@ -67,6 +67,24 @@
...
+    ansible.posix.seboolean:
+      name: antivirus_can_scan_system
+      state: yes
+      persistent: yes
```
Comment (jsf9k) [inline tasks/main.yml]: Alphabetize:
```suggestion
      name: antivirus_can_scan_system
      persistent: yes
      state: yes
```

**Review by jsf9k — CHANGES_REQUESTED** [review]
More minor things.

File: tasks/main.yml (hunk 3 above)
Comment (jsf9k) [reply tasks/main.yml]: Resolved via commit f34d0bb.

**Comment by jsf9k** [comment]
I spent a lot of time thinking about how to make the SELinux policy changes that you are adding in a clean way that passes `ansible-lint`.  After that I started to wonder why this change is necessary in the first place.

The problem is that [the SELinux policy for `freshclam`](https://github.com/fedora-selinux/selinux-policy/blob/f37/policy/modules/contrib/clamav.te) is [only allowing the standard HTTP port](https://github.com/fedora-selinux/selinux-policy/blob/f37/policy/modules/contrib/clamav.te#L232-L234).  You are trying to connect to a private mirror on port `8000`, which is in violation of this policy.

I think the answer here is to either 
- Make the changes to the SELinux policy that are required for your local environment (i.e., allow `freshclam` to reach out on port `8000`) _outside_ of this Ansible role
- Or, have your local ClamAV mirror start listening on port `80`

In any event, I don't think these changes belong in this Ansible role.  This Ansible role is intended to install and configure ClamAV, but it cannot support generic tweaking of the SELinux policy for your local environment; such support belongs, perhaps, in a locale-specific Ansible role.

If you feel differently, feel free to fork this repo and create your own version of this Ansible tole.

**Comment by mjourdan** [comment]
Having clamav to work with a private mirror listening on its default port sounded like a common-enough use-case to me, so I figured I would share my changes within this role.

Also there is a second part to this PR, where we [allow clamav to access the system](https://www.systutorials.com/docs/linux/man/8-antivirus_selinux/), which might be required on RedHat variants [according to clamavAV documentation](https://docs.clamav.net/manual/Usage/Configuration.html#configure-selinux-for-clamav).

Until we figure out a better approach, I think we'll keep using our fork for these tweaks. Anyway, thank you for the time you spent on this!

**Comment by jsf9k** [comment]
> Also there is a second part to this PR, where we [allow clamav to access the system](https://www.systutorials.com/docs/linux/man/8-antivirus_selinux/), which might be required on RedHat variants [according to clamavAV documentation](https://docs.clamav.net/manual/Usage/Configuration.html#configure-selinux-for-clamav).

If you want to create a new PR that only includes this change that would be something we would include.

**Pull request closed without merging**